// react-day-picker/dist/esm/helpers/getLabels.js
import * as d from "../labels/index.js";
import { labelWeekday as p } from "../labels/labelWeekday.js";
import { labelWeekNumberHeader as f } from "../labels/labelWeekNumberHeader.js";
import { labelNav as N } from "../labels/labelNav.js";
import { labelGridcell as k } from "../labels/labelGridcell.js";
import { labelGrid as D } from "../labels/labelGrid.js";
import { labelYearDropdown as W } from "../labels/labelYearDropdown.js";
import { labelWeekNumber as a } from "../labels/labelWeekNumber.js";
import { labelPrevious as v } from "../labels/labelPrevious.js";
import { labelNext as y } from "../labels/labelNext.js";
import { labelMonthDropdown as w } from "../labels/labelMonthDropdown.js";
import { labelDayButton as G } from "../labels/labelDayButton.js";
const l = (r, i, e) => i || (e ? typeof e == "function" ? e : (...n) => e : r);
function q(r, i) {
  var n;
  const e = ((n = i.locale) == null ? void 0 : n.labels) ?? {};
  return {
    ...d,
    ...r ?? {},
    labelDayButton: l(G, r == null ? void 0 : r.labelDayButton, e.labelDayButton),
    labelMonthDropdown: l(w, r == null ? void 0 : r.labelMonthDropdown, e.labelMonthDropdown),
    labelNext: l(y, r == null ? void 0 : r.labelNext, e.labelNext),
    labelPrevious: l(v, r == null ? void 0 : r.labelPrevious, e.labelPrevious),
    labelWeekNumber: l(a, r == null ? void 0 : r.labelWeekNumber, e.labelWeekNumber),
    labelYearDropdown: l(W, r == null ? void 0 : r.labelYearDropdown, e.labelYearDropdown),
    labelGrid: l(D, r == null ? void 0 : r.labelGrid, e.labelGrid),
    labelGridcell: l(k, r == null ? void 0 : r.labelGridcell, e.labelGridcell),
    labelNav: l(N, r == null ? void 0 : r.labelNav, e.labelNav),
    labelWeekNumberHeader: l(f, r == null ? void 0 : r.labelWeekNumberHeader, e.labelWeekNumberHeader),
    labelWeekday: l(p, r == null ? void 0 : r.labelWeekday, e.labelWeekday)
  };
}
export {
  q as getLabels
};
