// react-day-picker/dist/esm/DayPicker.js
import m, { useMemo as U, useCallback as h, useRef as Ie } from "react";
import { DateLib as Ke } from "./classes/DateLib.js";
import { createGetModifiers as Ge } from "./helpers/createGetModifiers.js";
import { getClassNamesForModifiers as He } from "./helpers/getClassNamesForModifiers.js";
import { getComponents as _e } from "./helpers/getComponents.js";
import { getDataAttributes as Ue } from "./helpers/getDataAttributes.js";
import { getDefaultClassNames as je } from "./helpers/getDefaultClassNames.js";
import { getFormatters as Ve } from "./helpers/getFormatters.js";
import { getLabels as $e } from "./helpers/getLabels.js";
import { getMonthOptions as qe } from "./helpers/getMonthOptions.js";
import { getStyleForModifiers as ze } from "./helpers/getStyleForModifiers.js";
import { getWeekdays as Je } from "./helpers/getWeekdays.js";
import { getYearOptions as Qe } from "./helpers/getYearOptions.js";
import { createNoonOverrides as Xe } from "./noonDateLib.js";
import { UI as r, DayFlag as et, SelectionState as P } from "./UI.js";
import { useAnimation as tt } from "./useAnimation.js";
import { useCalendar as at } from "./useCalendar.js";
import { dayPickerContext as ot } from "./useDayPicker.js";
import { useFocus as nt } from "./useFocus.js";
import { useSelection as rt } from "./useSelection.js";
import { convertMatchersToTimeZone as j } from "./utils/convertMatchersToTimeZone.js";
import { rangeIncludesDate as it } from "./utils/rangeIncludesDate.js";
import { toTimeZone as k } from "./utils/toTimeZone.js";
import { isDateRange as dt } from "./utils/typeguards.js";
import { enUS as st } from "./locale/en-US.js";
function Lt(V) {
  var ie;
  let e = V;
  const f = e.timeZone;
  if (f && (e = {
    ...V,
    timeZone: f
  }, e.today && (e.today = k(e.today, f)), e.month && (e.month = k(e.month, f)), e.defaultMonth && (e.defaultMonth = k(e.defaultMonth, f)), e.startMonth && (e.startMonth = k(e.startMonth, f)), e.endMonth && (e.endMonth = k(e.endMonth, f)), e.mode === "single" && e.selected ? e.selected = k(e.selected, f) : e.mode === "multiple" && e.selected ? e.selected = (ie = e.selected) == null ? void 0 : ie.map((n) => k(n, f)) : e.mode === "range" && e.selected && (e.selected = {
    from: e.selected.from ? k(e.selected.from, f) : e.selected.from,
    to: e.selected.to ? k(e.selected.to, f) : e.selected.to
  }), e.disabled !== void 0 && (e.disabled = j(e.disabled, f)), e.hidden !== void 0 && (e.hidden = j(e.hidden, f)), e.modifiers)) {
    const n = {};
    Object.keys(e.modifiers).forEach((s) => {
      var a;
      n[s] = j((a = e.modifiers) == null ? void 0 : a[s], f);
    }), e.modifiers = n;
  }
  const { components: c, formatters: D, labels: $, dateLib: o, locale: q, classNames: l } = U(() => {
    const n = { ...st, ...e.locale }, s = e.broadcastCalendar ? 1 : e.weekStartsOn, a = e.noonSafe && e.timeZone ? Xe(e.timeZone, {
      weekStartsOn: s,
      locale: n
    }) : void 0, d = e.dateLib && a ? { ...a, ...e.dateLib } : e.dateLib ?? a, u = new Ke({
      locale: n,
      weekStartsOn: s,
      firstWeekContainsDate: e.firstWeekContainsDate,
      useAdditionalWeekYearTokens: e.useAdditionalWeekYearTokens,
      useAdditionalDayOfYearTokens: e.useAdditionalDayOfYearTokens,
      timeZone: e.timeZone,
      numerals: e.numerals
    }, d);
    return {
      dateLib: u,
      components: _e(e.components),
      formatters: Ve(e.formatters),
      labels: $e(e.labels, u.options),
      locale: n,
      classNames: { ...je(), ...e.classNames }
    };
  }, [
    e.locale,
    e.broadcastCalendar,
    e.weekStartsOn,
    e.firstWeekContainsDate,
    e.useAdditionalWeekYearTokens,
    e.useAdditionalDayOfYearTokens,
    e.timeZone,
    e.numerals,
    e.dateLib,
    e.noonSafe,
    e.components,
    e.formatters,
    e.labels,
    e.classNames
  ]);
  e.today || (e = { ...e, today: o.today() });
  const { captionLayout: v, mode: F, navLayout: w, numberOfMonths: z = 1, onDayBlur: W, onDayClick: g, onDayFocus: C, onDayKeyDown: E, onDayMouseEnter: O, onDayMouseLeave: B, onNextClick: S, onPrevClick: Y, showWeekNumber: J, styles: t } = e, { formatCaption: Q, formatDay: X, formatMonthDropdown: de, formatWeekNumber: se, formatWeekNumberHeader: le, formatWeekdayName: me, formatYearDropdown: ce } = D, ee = at(e, o), { days: fe, months: T, navStart: R, navEnd: Z, previousMonth: p, nextMonth: b, goToMonth: N } = ee, I = Ge(fe, e, R, Z, o), { isSelected: M, select: y, selected: x } = rt(e, o) ?? {}, { blur: te, focused: A, isFocusTarget: ue, moveFocus: ae, setFocused: L } = nt(e, ee, I, M ?? (() => !1), o), { labelDayButton: pe, labelGridcell: be, labelGrid: he, labelMonthDropdown: Ne, labelNav: oe, labelPrevious: ke, labelNext: ve, labelWeekday: ge, labelWeekNumber: Me, labelWeekNumberHeader: ye, labelYearDropdown: De } = $, we = U(() => Je(o, e.ISOWeek, e.broadcastCalendar, e.today), [o, e.ISOWeek, e.broadcastCalendar, e.today]), ne = F !== void 0 || g !== void 0, K = h(() => {
    p && (N(p), Y == null || Y(p));
  }, [p, N, Y]), G = h(() => {
    b && (N(b), S == null || S(b));
  }, [N, b, S]), We = h((n, s) => (a) => {
    a.preventDefault(), a.stopPropagation(), L(n), !s.disabled && (y == null || y(n.date, s, a), g == null || g(n.date, s, a));
  }, [y, g, L]), Ce = h((n, s) => (a) => {
    L(n), C == null || C(n.date, s, a);
  }, [C, L]), Ee = h((n, s) => (a) => {
    te(), W == null || W(n.date, s, a);
  }, [te, W]), Oe = h((n, s) => (a) => {
    const d = {
      ArrowLeft: [
        a.shiftKey ? "month" : "day",
        e.dir === "rtl" ? "after" : "before"
      ],
      ArrowRight: [
        a.shiftKey ? "month" : "day",
        e.dir === "rtl" ? "before" : "after"
      ],
      ArrowDown: [a.shiftKey ? "year" : "week", "after"],
      ArrowUp: [a.shiftKey ? "year" : "week", "before"],
      PageUp: [a.shiftKey ? "year" : "month", "before"],
      PageDown: [a.shiftKey ? "year" : "month", "after"],
      Home: ["startOfWeek", "before"],
      End: ["endOfWeek", "after"]
    };
    if (d[a.key]) {
      a.preventDefault(), a.stopPropagation();
      const [u, i] = d[a.key];
      ae(u, i);
    }
    E == null || E(n.date, s, a);
  }, [ae, E, e.dir]), Be = h((n, s) => (a) => {
    O == null || O(n.date, s, a);
  }, [O]), Se = h((n, s) => (a) => {
    B == null || B(n.date, s, a);
  }, [B]), Ye = h((n) => (s) => {
    const a = Number(s.target.value), d = o.setMonth(o.startOfMonth(n), a);
    N(d);
  }, [o, N]), xe = h((n) => (s) => {
    const a = Number(s.target.value), d = o.setYear(o.startOfMonth(n), a);
    N(d);
  }, [o, N]), { className: Ae, style: Le } = U(() => ({
    className: [l[r.Root], e.className].filter(Boolean).join(" "),
    style: { ...t == null ? void 0 : t[r.Root], ...e.style }
  }), [l, e.className, e.style, t]), Pe = Ue(e), re = Ie(null);
  tt(re, !!e.animate, {
    classNames: l,
    months: T,
    focused: A,
    dateLib: o
  });
  const Fe = {
    dayPickerProps: e,
    selected: x,
    select: y,
    isSelected: M,
    months: T,
    nextMonth: b,
    previousMonth: p,
    goToMonth: N,
    getModifiers: I,
    components: c,
    classNames: l,
    styles: t,
    labels: $,
    formatters: D
  };
  return m.createElement(
    ot.Provider,
    { value: Fe },
    m.createElement(
      c.Root,
      { rootRef: e.animate ? re : void 0, className: Ae, style: Le, dir: e.dir, id: e.id, lang: e.lang ?? q.code, nonce: e.nonce, title: e.title, role: e.role, "aria-label": e["aria-label"], "aria-labelledby": e["aria-labelledby"], ...Pe },
      m.createElement(
        c.Months,
        { className: l[r.Months], style: t == null ? void 0 : t[r.Months] },
        !e.hideNavigation && !w && m.createElement(c.Nav, { "data-animated-nav": e.animate ? "true" : void 0, className: l[r.Nav], style: t == null ? void 0 : t[r.Nav], "aria-label": oe(), onPreviousClick: K, onNextClick: G, previousMonth: p, nextMonth: b }),
        T.map((n, s) => m.createElement(
          c.Month,
          {
            "data-animated-month": e.animate ? "true" : void 0,
            className: l[r.Month],
            style: t == null ? void 0 : t[r.Month],
            // biome-ignore lint/suspicious/noArrayIndexKey: breaks animation
            key: s,
            displayIndex: s,
            calendarMonth: n
          },
          w === "around" && !e.hideNavigation && s === 0 && m.createElement(
            c.PreviousMonthButton,
            { type: "button", className: l[r.PreviousMonthButton], tabIndex: p ? void 0 : -1, "aria-disabled": p ? void 0 : !0, "aria-label": ke(p), onClick: K, "data-animated-button": e.animate ? "true" : void 0 },
            m.createElement(c.Chevron, { disabled: p ? void 0 : !0, className: l[r.Chevron], orientation: e.dir === "rtl" ? "right" : "left" })
          ),
          m.createElement(c.MonthCaption, { "data-animated-caption": e.animate ? "true" : void 0, className: l[r.MonthCaption], style: t == null ? void 0 : t[r.MonthCaption], calendarMonth: n, displayIndex: s }, v != null && v.startsWith("dropdown") ? m.createElement(
            c.DropdownNav,
            { className: l[r.Dropdowns], style: t == null ? void 0 : t[r.Dropdowns] },
            (() => {
              const a = v === "dropdown" || v === "dropdown-months" ? m.createElement(c.MonthsDropdown, { key: "month", className: l[r.MonthsDropdown], "aria-label": Ne(), classNames: l, components: c, disabled: !!e.disableNavigation, onChange: Ye(n.date), options: qe(n.date, R, Z, D, o), style: t == null ? void 0 : t[r.Dropdown], value: o.getMonth(n.date) }) : m.createElement("span", { key: "month" }, de(n.date, o)), d = v === "dropdown" || v === "dropdown-years" ? m.createElement(c.YearsDropdown, { key: "year", className: l[r.YearsDropdown], "aria-label": De(o.options), classNames: l, components: c, disabled: !!e.disableNavigation, onChange: xe(n.date), options: Qe(R, Z, D, o, !!e.reverseYears), style: t == null ? void 0 : t[r.Dropdown], value: o.getYear(n.date) }) : m.createElement("span", { key: "year" }, ce(n.date, o));
              return o.getMonthYearOrder() === "year-first" ? [d, a] : [a, d];
            })(),
            m.createElement("span", { role: "status", "aria-live": "polite", style: {
              border: 0,
              clip: "rect(0 0 0 0)",
              height: "1px",
              margin: "-1px",
              overflow: "hidden",
              padding: 0,
              position: "absolute",
              width: "1px",
              whiteSpace: "nowrap",
              wordWrap: "normal"
            } }, Q(n.date, o.options, o))
          ) : m.createElement(c.CaptionLabel, { className: l[r.CaptionLabel], role: "status", "aria-live": "polite" }, Q(n.date, o.options, o))),
          w === "around" && !e.hideNavigation && s === z - 1 && m.createElement(
            c.NextMonthButton,
            { type: "button", className: l[r.NextMonthButton], tabIndex: b ? void 0 : -1, "aria-disabled": b ? void 0 : !0, "aria-label": ve(b), onClick: G, "data-animated-button": e.animate ? "true" : void 0 },
            m.createElement(c.Chevron, { disabled: b ? void 0 : !0, className: l[r.Chevron], orientation: e.dir === "rtl" ? "left" : "right" })
          ),
          s === z - 1 && w === "after" && !e.hideNavigation && m.createElement(c.Nav, { "data-animated-nav": e.animate ? "true" : void 0, className: l[r.Nav], style: t == null ? void 0 : t[r.Nav], "aria-label": oe(), onPreviousClick: K, onNextClick: G, previousMonth: p, nextMonth: b }),
          m.createElement(
            c.MonthGrid,
            { role: "grid", "aria-multiselectable": F === "multiple" || F === "range", "aria-label": he(n.date, o.options, o) || void 0, className: l[r.MonthGrid], style: t == null ? void 0 : t[r.MonthGrid] },
            !e.hideWeekdays && m.createElement(
              c.Weekdays,
              { "data-animated-weekdays": e.animate ? "true" : void 0, className: l[r.Weekdays], style: t == null ? void 0 : t[r.Weekdays] },
              J && m.createElement(c.WeekNumberHeader, { "aria-label": ye(o.options), className: l[r.WeekNumberHeader], style: t == null ? void 0 : t[r.WeekNumberHeader], scope: "col" }, le()),
              we.map((a) => m.createElement(c.Weekday, { "aria-label": ge(a, o.options, o), className: l[r.Weekday], key: String(a), style: t == null ? void 0 : t[r.Weekday], scope: "col" }, me(a, o.options, o)))
            ),
            m.createElement(c.Weeks, { "data-animated-weeks": e.animate ? "true" : void 0, className: l[r.Weeks], style: t == null ? void 0 : t[r.Weeks] }, n.weeks.map((a) => m.createElement(
              c.Week,
              { className: l[r.Week], key: a.weekNumber, style: t == null ? void 0 : t[r.Week], week: a },
              J && m.createElement(c.WeekNumber, { week: a, style: t == null ? void 0 : t[r.WeekNumber], "aria-label": Me(a.weekNumber, {
                locale: q
              }), className: l[r.WeekNumber], scope: "row", role: "rowheader" }, se(a.weekNumber, o)),
              a.days.map((d) => {
                const { date: u } = d, i = I(d);
                if (i[et.focused] = !i.hidden && !!(A != null && A.isEqualTo(d)), i[P.selected] = (M == null ? void 0 : M(u)) || i.selected, dt(x)) {
                  const { from: H, to: _ } = x;
                  i[P.range_start] = !!(H && _ && o.isSameDay(u, H)), i[P.range_end] = !!(H && _ && o.isSameDay(u, _)), i[P.range_middle] = it(x, u, !0, o);
                }
                const Te = ze(i, t, e.modifiersStyles), Re = He(i, l, e.modifiersClassNames), Ze = !ne && !i.hidden ? be(u, i, o.options, o) : void 0;
                return m.createElement(c.Day, { key: `${d.isoDate}_${d.displayMonthId}`, day: d, modifiers: i, className: Re.join(" "), style: Te, role: "gridcell", "aria-selected": i.selected || void 0, "aria-label": Ze, "data-day": d.isoDate, "data-month": d.outside ? d.dateMonthId : void 0, "data-selected": i.selected || void 0, "data-disabled": i.disabled || void 0, "data-hidden": i.hidden || void 0, "data-outside": d.outside || void 0, "data-focused": i.focused || void 0, "data-today": i.today || void 0 }, !i.hidden && ne ? m.createElement(c.DayButton, { className: l[r.DayButton], style: t == null ? void 0 : t[r.DayButton], type: "button", day: d, modifiers: i, disabled: !i.focused && i.disabled || void 0, "aria-disabled": i.focused && i.disabled || void 0, tabIndex: ue(d) ? 0 : -1, "aria-label": pe(u, i, o.options, o), onClick: We(d, i), onBlur: Ee(d, i), onFocus: Ce(d, i), onKeyDown: Oe(d, i), onMouseEnter: Be(d, i), onMouseLeave: Se(d, i) }, X(u, o.options, o)) : !i.hidden && X(d.date, o.options, o));
              })
            )))
          )
        ))
      ),
      e.footer && m.createElement(c.Footer, { className: l[r.Footer], style: t == null ? void 0 : t[r.Footer], role: "status", "aria-live": "polite" }, e.footer)
    )
  );
}
export {
  Lt as DayPicker
};
