// react-day-picker/dist/esm/useCalendar.js
import { useEffect as N, useMemo as w } from "react";
import { getDates as S } from "./helpers/getDates.js";
import { getDays as b } from "./helpers/getDays.js";
import { getDisplayMonths as q } from "./helpers/getDisplayMonths.js";
import { getInitialMonth as g } from "./helpers/getInitialMonth.js";
import { getMonths as E } from "./helpers/getMonths.js";
import { getNavMonths as F } from "./helpers/getNavMonth.js";
import { getNextMonth as P } from "./helpers/getNextMonth.js";
import { getPreviousMonth as V } from "./helpers/getPreviousMonth.js";
import { getWeeks as Z } from "./helpers/getWeeks.js";
import { useControlledValue as j } from "./helpers/useControlledValue.js";
function $(e, o) {
  var l;
  const [s, t] = F(e, o), { startOfMonth: a, endOfMonth: u } = o, h = g(e, s, t, o), [r, M] = j(
    h,
    // initialMonth is always computed from props.month if provided
    e.month ? h : void 0
  );
  N(() => {
    const n = g(e, s, t, o);
    M(n);
  }, [e.timeZone]);
  const { months: k, weeks: c, days: W, previousMonth: O, nextMonth: x } = w(() => {
    const n = q(r, t, { numberOfMonths: e.numberOfMonths }, o), i = S(n, e.endMonth ? u(e.endMonth) : void 0, {
      ISOWeek: e.ISOWeek,
      fixedWeeks: e.fixedWeeks,
      broadcastCalendar: e.broadcastCalendar
    }, o), m = E(n, i, {
      broadcastCalendar: e.broadcastCalendar,
      fixedWeeks: e.fixedWeeks,
      ISOWeek: e.ISOWeek,
      reverseMonths: e.reverseMonths
    }, o), I = Z(m), T = b(m), C = V(r, s, e, o), D = P(r, t, e, o);
    return {
      months: m,
      weeks: I,
      days: T,
      previousMonth: C,
      nextMonth: D
    };
  }, [
    o,
    r.getTime(),
    t == null ? void 0 : t.getTime(),
    s == null ? void 0 : s.getTime(),
    e.disableNavigation,
    e.broadcastCalendar,
    (l = e.endMonth) == null ? void 0 : l.getTime(),
    e.fixedWeeks,
    e.ISOWeek,
    e.numberOfMonths,
    e.pagedNavigation,
    e.reverseMonths
  ]), { disableNavigation: v, onMonthChange: f } = e, y = (n) => c.some((i) => i.days.some((m) => m.isEqualTo(n))), d = (n) => {
    if (v)
      return;
    let i = a(n);
    s && i < a(s) && (i = a(s)), t && i > a(t) && (i = a(t)), M(i), f == null || f(i);
  };
  return {
    months: k,
    weeks: c,
    days: W,
    navStart: s,
    navEnd: t,
    previousMonth: O,
    nextMonth: x,
    goToMonth: d,
    goToDay: (n) => {
      y(n) || d(n.date);
    }
  };
}
export {
  $ as useCalendar
};
