// @radix-ui/react-tooltip/dist/index.js
import * as s from "react";
import { composeEventHandlers as g } from "../../primitive/dist/index.js";
import { useComposedRefs as N } from "../../react-compose-refs/dist/index.js";
import { createContextScope as W } from "../../react-context/dist/index.js";
import { DismissableLayer as $ } from "../../react-dismissable-layer/dist/index.js";
import { useId as z } from "../../react-id/dist/index.js";
import { Root as J, Arrow as Q, Anchor as Z, createPopperScope as S, Content as ee } from "../../react-popper/dist/index.js";
import { Portal as te } from "../node_modules/@radix-ui/react-portal/dist/index.js";
import { Presence as G } from "../../react-presence/dist/index.js";
import { Primitive as oe } from "../../react-primitive/dist/index.js";
import { createSlottable as re } from "../node_modules/@radix-ui/react-slot/dist/index.js";
import { useControllableState as ne } from "../../react-use-controllable-state/dist/index.js";
import { Root as se } from "../../react-visually-hidden/dist/index.js";
import { jsx as f, jsxs as ie } from "react/jsx-runtime";
var [A] = W("Tooltip", [
  S
]), D = S(), j = "TooltipProvider", ae = 700, O = "tooltip.open", [le, k] = A(j), F = (t) => {
  const {
    __scopeTooltip: o,
    delayDuration: e = ae,
    skipDelayDuration: r = 300,
    disableHoverableContent: n = !1,
    children: a
  } = t, l = s.useRef(!0), v = s.useRef(!1), i = s.useRef(0);
  return s.useEffect(() => {
    const p = i.current;
    return () => window.clearTimeout(p);
  }, []), /* @__PURE__ */ f(
    le,
    {
      scope: o,
      isOpenDelayedRef: l,
      delayDuration: e,
      onOpen: s.useCallback(() => {
        window.clearTimeout(i.current), l.current = !1;
      }, []),
      onClose: s.useCallback(() => {
        window.clearTimeout(i.current), i.current = window.setTimeout(
          () => l.current = !0,
          r
        );
      }, [r]),
      isPointerInTransitRef: v,
      onPointerInTransitChange: s.useCallback((p) => {
        v.current = p;
      }, []),
      disableHoverableContent: n,
      children: a
    }
  );
};
F.displayName = j;
var R = "Tooltip", [ce, _] = A(R), B = (t) => {
  const {
    __scopeTooltip: o,
    children: e,
    open: r,
    defaultOpen: n,
    onOpenChange: a,
    disableHoverableContent: l,
    delayDuration: v
  } = t, i = k(R, t.__scopeTooltip), p = D(o), [c, d] = s.useState(null), h = z(), u = s.useRef(0), m = l ?? i.disableHoverableContent, y = v ?? i.delayDuration, T = s.useRef(!1), [x, C] = ne({
    prop: r,
    defaultProp: n ?? !1,
    onChange: (H) => {
      H ? (i.onOpen(), document.dispatchEvent(new CustomEvent(O))) : i.onClose(), a == null || a(H);
    },
    caller: R
  }), w = s.useMemo(() => x ? T.current ? "delayed-open" : "instant-open" : "closed", [x]), P = s.useCallback(() => {
    window.clearTimeout(u.current), u.current = 0, T.current = !1, C(!0);
  }, [C]), E = s.useCallback(() => {
    window.clearTimeout(u.current), u.current = 0, C(!1);
  }, [C]), I = s.useCallback(() => {
    window.clearTimeout(u.current), u.current = window.setTimeout(() => {
      T.current = !0, C(!0), u.current = 0;
    }, y);
  }, [y, C]);
  return s.useEffect(() => () => {
    u.current && (window.clearTimeout(u.current), u.current = 0);
  }, []), /* @__PURE__ */ f(J, { ...p, children: /* @__PURE__ */ f(
    ce,
    {
      scope: o,
      contentId: h,
      open: x,
      stateAttribute: w,
      trigger: c,
      onTriggerChange: d,
      onTriggerEnter: s.useCallback(() => {
        i.isOpenDelayedRef.current ? I() : P();
      }, [i.isOpenDelayedRef, I, P]),
      onTriggerLeave: s.useCallback(() => {
        m ? E() : (window.clearTimeout(u.current), u.current = 0);
      }, [E, m]),
      onOpen: P,
      onClose: E,
      disableHoverableContent: m,
      children: e
    }
  ) });
};
B.displayName = R;
var L = "TooltipTrigger", U = s.forwardRef(
  (t, o) => {
    const { __scopeTooltip: e, ...r } = t, n = _(L, e), a = k(L, e), l = D(e), v = s.useRef(null), i = N(o, v, n.onTriggerChange), p = s.useRef(!1), c = s.useRef(!1), d = s.useCallback(() => p.current = !1, []);
    return s.useEffect(() => () => document.removeEventListener("pointerup", d), [d]), /* @__PURE__ */ f(Z, { asChild: !0, ...l, children: /* @__PURE__ */ f(
      oe.button,
      {
        "aria-describedby": n.open ? n.contentId : void 0,
        "data-state": n.stateAttribute,
        ...r,
        ref: i,
        onPointerMove: g(t.onPointerMove, (h) => {
          h.pointerType !== "touch" && !c.current && !a.isPointerInTransitRef.current && (n.onTriggerEnter(), c.current = !0);
        }),
        onPointerLeave: g(t.onPointerLeave, () => {
          n.onTriggerLeave(), c.current = !1;
        }),
        onPointerDown: g(t.onPointerDown, () => {
          n.open && n.onClose(), p.current = !0, document.addEventListener("pointerup", d, { once: !0 });
        }),
        onFocus: g(t.onFocus, () => {
          p.current || n.onOpen();
        }),
        onBlur: g(t.onBlur, n.onClose),
        onClick: g(t.onClick, n.onClose)
      }
    ) });
  }
);
U.displayName = L;
var M = "TooltipPortal", [ue, pe] = A(M, {
  forceMount: void 0
}), V = (t) => {
  const { __scopeTooltip: o, forceMount: e, children: r, container: n } = t, a = _(M, o);
  return /* @__PURE__ */ f(ue, { scope: o, forceMount: e, children: /* @__PURE__ */ f(G, { present: e || a.open, children: /* @__PURE__ */ f(te, { asChild: !0, container: n, children: r }) }) });
};
V.displayName = M;
var b = "TooltipContent", Y = s.forwardRef(
  (t, o) => {
    const e = pe(b, t.__scopeTooltip), { forceMount: r = e.forceMount, side: n = "top", ...a } = t, l = _(b, t.__scopeTooltip);
    return /* @__PURE__ */ f(G, { present: r || l.open, children: l.disableHoverableContent ? /* @__PURE__ */ f(q, { side: n, ...a, ref: o }) : /* @__PURE__ */ f(de, { side: n, ...a, ref: o }) });
  }
), de = s.forwardRef((t, o) => {
  const e = _(b, t.__scopeTooltip), r = k(b, t.__scopeTooltip), n = s.useRef(null), a = N(o, n), [l, v] = s.useState(null), { trigger: i, onClose: p } = e, c = n.current, { onPointerInTransitChange: d } = r, h = s.useCallback(() => {
    v(null), d(!1);
  }, [d]), u = s.useCallback(
    (m, y) => {
      const T = m.currentTarget, x = { x: m.clientX, y: m.clientY }, C = me(x, T.getBoundingClientRect()), w = Te(x, C), P = ye(y.getBoundingClientRect()), E = xe([...w, ...P]);
      v(E), d(!0);
    },
    [d]
  );
  return s.useEffect(() => () => h(), [h]), s.useEffect(() => {
    if (i && c) {
      const m = (T) => u(T, c), y = (T) => u(T, i);
      return i.addEventListener("pointerleave", m), c.addEventListener("pointerleave", y), () => {
        i.removeEventListener("pointerleave", m), c.removeEventListener("pointerleave", y);
      };
    }
  }, [i, c, u, h]), s.useEffect(() => {
    if (l) {
      const m = (y) => {
        const T = y.target, x = { x: y.clientX, y: y.clientY }, C = (i == null ? void 0 : i.contains(T)) || (c == null ? void 0 : c.contains(T)), w = !Ce(x, l);
        C ? h() : w && (h(), p());
      };
      return document.addEventListener("pointermove", m), () => document.removeEventListener("pointermove", m);
    }
  }, [i, c, l, p, h]), /* @__PURE__ */ f(q, { ...t, ref: a });
}), [fe, ve] = A(R, { isInside: !1 }), he = re("TooltipContent"), q = s.forwardRef(
  (t, o) => {
    const {
      __scopeTooltip: e,
      children: r,
      "aria-label": n,
      onEscapeKeyDown: a,
      onPointerDownOutside: l,
      ...v
    } = t, i = _(b, e), p = D(e), { onClose: c } = i;
    return s.useEffect(() => (document.addEventListener(O, c), () => document.removeEventListener(O, c)), [c]), s.useEffect(() => {
      if (i.trigger) {
        const d = (h) => {
          const u = h.target;
          u != null && u.contains(i.trigger) && c();
        };
        return window.addEventListener("scroll", d, { capture: !0 }), () => window.removeEventListener("scroll", d, { capture: !0 });
      }
    }, [i.trigger, c]), /* @__PURE__ */ f(
      $,
      {
        asChild: !0,
        disableOutsidePointerEvents: !1,
        onEscapeKeyDown: a,
        onPointerDownOutside: l,
        onFocusOutside: (d) => d.preventDefault(),
        onDismiss: c,
        children: /* @__PURE__ */ ie(
          ee,
          {
            "data-state": i.stateAttribute,
            ...p,
            ...v,
            ref: o,
            style: {
              ...v.style,
              "--radix-tooltip-content-transform-origin": "var(--radix-popper-transform-origin)",
              "--radix-tooltip-content-available-width": "var(--radix-popper-available-width)",
              "--radix-tooltip-content-available-height": "var(--radix-popper-available-height)",
              "--radix-tooltip-trigger-width": "var(--radix-popper-anchor-width)",
              "--radix-tooltip-trigger-height": "var(--radix-popper-anchor-height)"
            },
            children: [
              /* @__PURE__ */ f(he, { children: r }),
              /* @__PURE__ */ f(fe, { scope: e, isInside: !0, children: /* @__PURE__ */ f(se, { id: i.contentId, role: "tooltip", children: n || r }) })
            ]
          }
        )
      }
    );
  }
);
Y.displayName = b;
var X = "TooltipArrow", K = s.forwardRef(
  (t, o) => {
    const { __scopeTooltip: e, ...r } = t, n = D(e);
    return ve(
      X,
      e
    ).isInside ? null : /* @__PURE__ */ f(Q, { ...n, ...r, ref: o });
  }
);
K.displayName = X;
function me(t, o) {
  const e = Math.abs(o.top - t.y), r = Math.abs(o.bottom - t.y), n = Math.abs(o.right - t.x), a = Math.abs(o.left - t.x);
  switch (Math.min(e, r, n, a)) {
    case a:
      return "left";
    case n:
      return "right";
    case e:
      return "top";
    case r:
      return "bottom";
    default:
      throw new Error("unreachable");
  }
}
function Te(t, o, e = 5) {
  const r = [];
  switch (o) {
    case "top":
      r.push(
        { x: t.x - e, y: t.y + e },
        { x: t.x + e, y: t.y + e }
      );
      break;
    case "bottom":
      r.push(
        { x: t.x - e, y: t.y - e },
        { x: t.x + e, y: t.y - e }
      );
      break;
    case "left":
      r.push(
        { x: t.x + e, y: t.y - e },
        { x: t.x + e, y: t.y + e }
      );
      break;
    case "right":
      r.push(
        { x: t.x - e, y: t.y - e },
        { x: t.x - e, y: t.y + e }
      );
      break;
  }
  return r;
}
function ye(t) {
  const { top: o, right: e, bottom: r, left: n } = t;
  return [
    { x: n, y: o },
    { x: e, y: o },
    { x: e, y: r },
    { x: n, y: r }
  ];
}
function Ce(t, o) {
  const { x: e, y: r } = t;
  let n = !1;
  for (let a = 0, l = o.length - 1; a < o.length; l = a++) {
    const v = o[a], i = o[l], p = v.x, c = v.y, d = i.x, h = i.y;
    c > r != h > r && e < (d - p) * (r - c) / (h - c) + p && (n = !n);
  }
  return n;
}
function xe(t) {
  const o = t.slice();
  return o.sort((e, r) => e.x < r.x ? -1 : e.x > r.x ? 1 : e.y < r.y ? -1 : e.y > r.y ? 1 : 0), ge(o);
}
function ge(t) {
  if (t.length <= 1) return t.slice();
  const o = [];
  for (let r = 0; r < t.length; r++) {
    const n = t[r];
    for (; o.length >= 2; ) {
      const a = o[o.length - 1], l = o[o.length - 2];
      if ((a.x - l.x) * (n.y - l.y) >= (a.y - l.y) * (n.x - l.x)) o.pop();
      else break;
    }
    o.push(n);
  }
  o.pop();
  const e = [];
  for (let r = t.length - 1; r >= 0; r--) {
    const n = t[r];
    for (; e.length >= 2; ) {
      const a = e[e.length - 1], l = e[e.length - 2];
      if ((a.x - l.x) * (n.y - l.y) >= (a.y - l.y) * (n.x - l.x)) e.pop();
      else break;
    }
    e.push(n);
  }
  return e.pop(), o.length === 1 && e.length === 1 && o[0].x === e[0].x && o[0].y === e[0].y ? o : o.concat(e);
}
var He = F, Ne = B, Se = U, Ge = V, je = Y, Fe = K;
export {
  Fe as Arrow,
  je as Content,
  Ge as Portal,
  He as Provider,
  Ne as Root,
  B as Tooltip,
  K as TooltipArrow,
  Y as TooltipContent,
  V as TooltipPortal,
  F as TooltipProvider,
  U as TooltipTrigger,
  Se as Trigger
};
