// echarts/lib/core/echarts.js
import { __extends as Ee } from "tslib";
import { init as Ne, registerSSRDataGetter as Rt } from "../../../zrender/lib/zrender.js";
import { retrieve2 as X, bind as Ue, setAsPrimitive as Ct, isObject as B, isString as Ot, clone as ke, each as R, extend as z, assert as G, createHashMap as ue, map as Pt, defaults as be, indexOf as nt, isFunction as at, noop as K } from "../../../zrender/lib/core/util.js";
import q from "../../../zrender/lib/core/env.js";
import fe from "../../../zrender/lib/core/timsort.js";
import Oe from "../../../zrender/lib/core/Eventful.js";
import wt from "../model/Global.js";
import At from "./ExtensionAPI.js";
import st from "./CoordinateSystem.js";
import Dt from "../model/OptionManager.js";
import ot from "../preprocessor/backwardCompat.js";
import It from "../processor/dataStack.js";
import ze from "../model/Series.js";
import Vt from "../view/Component.js";
import de from "../view/Chart.js";
import { retrieveZInfo as Mt, traverseUpdateZ as Lt } from "../util/graphic.js";
import { getECData as $ } from "../util/innerStore.js";
import { isSelectChangePayload as Fe, isHighDownPayload as Ye, allLeaveBlur as xt, updateSeriesElementSelection as He, HOVER_STATE_EMPHASIS as Nt, HOVER_STATE_BLUR as Ut, savePathStates as kt, enterEmphasis as Be, leaveEmphasis as Ge, enterBlur as bt, leaveBlur as zt, enterSelect as Ft, leaveSelect as Yt, isHighDownDispatcher as Ze, handleGlobalMouseOverForHighDown as Ht, handleGlobalMouseOutForHighDown as Bt, HIGHLIGHT_ACTION_TYPE as Z, DOWNPLAY_ACTION_TYPE as pe, SELECT_ACTION_TYPE as We, SELECT_CHANGED_EVENT_TYPE as Pe, UNSELECT_ACTION_TYPE as Xe, TOGGLE_SELECT_ACTION_TYPE as qe, getAllSelectedIndices as Gt, blurSeriesFromHighlightPayload as Zt, findComponentHighDownDispatchers as Wt, blurComponent as Xt, toggleSelectionFromPayload as qt } from "../util/states.js";
import { parseFinder as le, setAttribute as Kt, preParseFinder as jt, convertOptionIdName as $t, normalizeToArray as Jt } from "../util/model.js";
import { throttle as Qt } from "../util/throttle.js";
import { seriesStyleTask as er, dataStyleTask as tr, dataColorPaletteTask as rr } from "../visual/style.js";
import ir from "../loading/default.js";
import ut from "./Scheduler.js";
import nr from "../theme/dark.js";
import { parseClassType as Ke } from "../util/clazz.js";
import { ECEventProcessor as ar } from "../util/ECEventProcessor.js";
import { seriesSymbolTask as sr, dataSymbolTask as or } from "../visual/symbol.js";
import { getItemVisualFromData as ur, getVisualFromData as fr } from "../visual/helper.js";
import { deprecateLog as dr, error as ie, deprecateReplaceLog as pr, warn as b } from "../util/log.js";
import { handleLegacySelectEvents as lr } from "../legacy/dataSelectAction.js";
import { registerExternalTransform as cr } from "../data/helper/transform.js";
import { createLocaleObject as hr, SYSTEM_LANG as vr } from "./locale.js";
import { registerLocale as Ui } from "./locale.js";
import { findEventDispatcher as J } from "../util/event.js";
import _r from "../visual/decal.js";
import L from "./lifecycle.js";
import { platformApi as gr } from "../../../zrender/lib/core/platform.js";
import { getImpl as mr } from "./impl.js";
import Er from "../../../zrender/lib/graphic/shape/Rect.js";
import Tr from "../../../zrender/lib/graphic/Path.js";
import Sr from "../../../zrender/lib/graphic/Image.js";
import { isElementRemoved as ce } from "../animation/basicTransition.js";
var yr = 1, Rr = 800, Cr = 900, Or = 1e3, Pr = 2e3, wr = 5e3, ft = 1e3, Ar = 1100, we = 2e3, dt = 3e3, Dr = 4e3, se = 4500, Ir = 4600, Vr = 5e3, Mr = 6e3, pt = 7e3, wi = {
  PROCESSOR: {
    FILTER: Or,
    SERIES_FILTER: Rr,
    STATISTIC: wr
  },
  VISUAL: {
    LAYOUT: ft,
    PROGRESSIVE_LAYOUT: Ar,
    GLOBAL: we,
    CHART: dt,
    POST_CHART_LAYOUT: Ir,
    COMPONENT: Dr,
    BRUSH: Vr,
    CHART_ITEM: se,
    ARIA: Mr,
    DECAL: pt
  }
}, C = "__flagInMainProcess", Q = "__mainProcessVersion", O = "__pendingUpdate", he = "__needsUpdateStatus", je = /^[a-zA-Z0-9_]+$/;
function lt(c) {
  return function() {
    for (var o = [], r = 0; r < arguments.length; r++)
      o[r] = arguments[r];
    if (this.isDisposed()) {
      D(this.id);
      return;
    }
    return ht(this, c, o);
  };
}
function ct(c) {
  return function() {
    for (var o = [], r = 0; r < arguments.length; r++)
      o[r] = arguments[r];
    return ht(this, c, o);
  };
}
function ht(c, o, r) {
  return r[0] = r[0] && r[0].toLowerCase(), Oe.prototype[o].apply(c, r);
}
var vt = (
  /** @class */
  (function(c) {
    Ee(o, c);
    function o() {
      return c !== null && c.apply(this, arguments) || this;
    }
    return o;
  })(Oe)
), _t = vt.prototype;
_t.on = ct("on");
_t.off = ct("off");
var k, ve, ee, x, te, _e, ge, F, Y, $e, Je, me, Qe, re, et, V, tt, H, Lr = (
  /** @class */
  (function(c) {
    Ee(o, c);
    function o(r, n, d) {
      var s = c.call(this, new ar()) || this;
      s._chartsViews = [], s._chartsMap = {}, s._componentsViews = [], s._componentsMap = {}, s._pendingActions = [], d = d || {}, s._dom = r;
      var g = "canvas", S = "auto", E = !1;
      if (s[Q] = 1, process.env.NODE_ENV !== "production") {
        var m = (
          /* eslint-disable-next-line */
          q.hasGlobalWindow ? window : global
        );
        m && (g = X(m.__ECHARTS__DEFAULT__RENDERER__, g), S = X(m.__ECHARTS__DEFAULT__COARSE_POINTER, S), E = X(m.__ECHARTS__DEFAULT__USE_DIRTY_RECT__, E));
      }
      d.ssr && Rt(function(t) {
        var a = $(t), u = a.dataIndex;
        if (u != null) {
          var p = ue();
          return p.set("series_index", a.seriesIndex), p.set("data_index", u), a.ssrType && p.set("ssr_type", a.ssrType), p;
        }
      });
      var y = s._zr = Ne(r, {
        renderer: d.renderer || g,
        devicePixelRatio: d.devicePixelRatio,
        width: d.width,
        height: d.height,
        ssr: d.ssr,
        useDirtyRect: X(d.useDirtyRect, E),
        useCoarsePointer: X(d.useCoarsePointer, S),
        pointerSize: d.pointerSize
      });
      s._ssr = d.ssr, s._throttledZrFlush = Qt(Ue(y.flush, y), 17), s._updateTheme(n), s._locale = hr(d.locale || vr), s._coordSysMgr = new st();
      var e = s._api = et(s);
      function i(t, a) {
        return t.__prio - a.__prio;
      }
      return fe(ae, i), fe(ye, i), s._scheduler = new ut(s, e, ye, ae), s._messageCenter = new vt(), s._initEvents(), s.resize = Ue(s.resize, s), y.animation.on("frame", s._onframe, s), $e(y, s), Je(y, s), Ct(s), s;
    }
    return o.prototype._onframe = function() {
      if (!this._disposed) {
        tt(this);
        var r = this._scheduler;
        if (this[O]) {
          var n = this[O].silent;
          this[C] = !0, H(this);
          try {
            k(this), x.update.call(this, null, this[O].updateParams);
          } catch (E) {
            throw this[C] = !1, this[O] = null, E;
          }
          this._zr.flush(), this[C] = !1, this[O] = null, F.call(this, n), Y.call(this, n);
        } else if (r.unfinished) {
          var d = yr, s = this._model, g = this._api;
          r.unfinished = !1;
          do {
            var S = +/* @__PURE__ */ new Date();
            r.performSeriesTasks(s), r.performDataProcessorTasks(s), _e(this, s), r.performVisualTasks(s), re(this, this._model, g, "remain", {}), d -= +/* @__PURE__ */ new Date() - S;
          } while (d > 0 && r.unfinished);
          r.unfinished || this._zr.flush();
        }
      }
    }, o.prototype.getDom = function() {
      return this._dom;
    }, o.prototype.getId = function() {
      return this.id;
    }, o.prototype.getZr = function() {
      return this._zr;
    }, o.prototype.isSSR = function() {
      return this._ssr;
    }, o.prototype.setOption = function(r, n, d) {
      if (this[C]) {
        process.env.NODE_ENV !== "production" && ie("`setOption` should not be called during main process.");
        return;
      }
      if (this._disposed) {
        D(this.id);
        return;
      }
      var s, g, S;
      if (B(n) && (d = n.lazyUpdate, s = n.silent, g = n.replaceMerge, S = n.transition, n = n.notMerge), this[C] = !0, H(this), !this._model || n) {
        var E = new Dt(this._api), m = this._theme, y = this._model = new wt();
        y.scheduler = this._scheduler, y.ssr = this._ssr, y.init(null, null, null, m, this._locale, E);
      }
      this._model.setOption(r, {
        replaceMerge: g
      }, Re);
      var e = {
        seriesTransition: S,
        optionChanged: !0
      };
      if (d)
        this[O] = {
          silent: s,
          updateParams: e
        }, this[C] = !1, this.getZr().wakeUp();
      else {
        try {
          k(this), x.update.call(this, null, e);
        } catch (i) {
          throw this[O] = null, this[C] = !1, i;
        }
        this._ssr || this._zr.flush(), this[O] = null, this[C] = !1, F.call(this, s), Y.call(this, s);
      }
    }, o.prototype.setTheme = function(r, n) {
      if (this[C]) {
        process.env.NODE_ENV !== "production" && ie("`setTheme` should not be called during main process.");
        return;
      }
      if (this._disposed) {
        D(this.id);
        return;
      }
      var d = this._model;
      if (d) {
        var s = n && n.silent, g = null;
        this[O] && (s == null && (s = this[O].silent), g = this[O].updateParams, this[O] = null), this[C] = !0, H(this);
        try {
          this._updateTheme(r), d.setTheme(this._theme), k(this), x.update.call(this, {
            type: "setTheme"
          }, g);
        } catch (S) {
          throw this[C] = !1, S;
        }
        this[C] = !1, F.call(this, s), Y.call(this, s);
      }
    }, o.prototype._updateTheme = function(r) {
      Ot(r) && (r = gt[r]), r && (r = ke(r), r && ot(r, !0), this._theme = r);
    }, o.prototype.getModel = function() {
      return this._model;
    }, o.prototype.getOption = function() {
      return this._model && this._model.getOption();
    }, o.prototype.getWidth = function() {
      return this._zr.getWidth();
    }, o.prototype.getHeight = function() {
      return this._zr.getHeight();
    }, o.prototype.getDevicePixelRatio = function() {
      return this._zr.painter.dpr || q.hasGlobalWindow && window.devicePixelRatio || 1;
    }, o.prototype.getRenderedCanvas = function(r) {
      return process.env.NODE_ENV !== "production" && pr("getRenderedCanvas", "renderToCanvas"), this.renderToCanvas(r);
    }, o.prototype.renderToCanvas = function(r) {
      r = r || {};
      var n = this._zr.painter;
      if (process.env.NODE_ENV !== "production" && n.type !== "canvas")
        throw new Error("renderToCanvas can only be used in the canvas renderer.");
      return n.getRenderedCanvas({
        backgroundColor: r.backgroundColor || this._model.get("backgroundColor"),
        pixelRatio: r.pixelRatio || this.getDevicePixelRatio()
      });
    }, o.prototype.renderToSVGString = function(r) {
      r = r || {};
      var n = this._zr.painter;
      if (process.env.NODE_ENV !== "production" && n.type !== "svg")
        throw new Error("renderToSVGString can only be used in the svg renderer.");
      return n.renderToString({
        useViewBox: r.useViewBox
      });
    }, o.prototype.getSvgDataURL = function() {
      var r = this._zr, n = r.storage.getDisplayList();
      return R(n, function(d) {
        d.stopAnimation(null, !0);
      }), r.painter.toDataURL();
    }, o.prototype.getDataURL = function(r) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      r = r || {};
      var n = r.excludeComponents, d = this._model, s = [], g = this;
      R(n, function(E) {
        d.eachComponent({
          mainType: E
        }, function(m) {
          var y = g._componentsMap[m.__viewId];
          y.group.ignore || (s.push(y), y.group.ignore = !0);
        });
      });
      var S = this._zr.painter.getType() === "svg" ? this.getSvgDataURL() : this.renderToCanvas(r).toDataURL("image/" + (r && r.type || "png"));
      return R(s, function(E) {
        E.group.ignore = !1;
      }), S;
    }, o.prototype.getConnectedDataURL = function(r) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      var n = r.type === "svg", d = this.group, s = Math.min, g = Math.max, S = 1 / 0;
      if (Nr[d]) {
        var E = S, m = S, y = -S, e = -S, i = [], t = r && r.pixelRatio || this.getDevicePixelRatio();
        R(rt, function(l, v) {
          if (l.group === d) {
            var T = n ? l.getZr().painter.getSvgDom().innerHTML : l.renderToCanvas(ke(r)), _ = l.getDom().getBoundingClientRect();
            E = s(_.left, E), m = s(_.top, m), y = g(_.right, y), e = g(_.bottom, e), i.push({
              dom: T,
              left: _.left,
              top: _.top
            });
          }
        }), E *= t, m *= t, y *= t, e *= t;
        var a = y - E, u = e - m, p = gr.createCanvas(), f = Ne(p, {
          renderer: n ? "svg" : "canvas"
        });
        if (f.resize({
          width: a,
          height: u
        }), n) {
          var h = "";
          return R(i, function(l) {
            var v = l.left - E, T = l.top - m;
            h += '<g transform="translate(' + v + "," + T + ')">' + l.dom + "</g>";
          }), f.painter.getSvgRoot().innerHTML = h, r.connectedBackgroundColor && f.painter.setBackgroundColor(r.connectedBackgroundColor), f.refreshImmediately(), f.painter.toDataURL();
        } else
          return r.connectedBackgroundColor && f.add(new Er({
            shape: {
              x: 0,
              y: 0,
              width: a,
              height: u
            },
            style: {
              fill: r.connectedBackgroundColor
            }
          })), R(i, function(l) {
            var v = new Sr({
              style: {
                x: l.left * t - E,
                y: l.top * t - m,
                image: l.dom
              }
            });
            f.add(v);
          }), f.refreshImmediately(), p.toDataURL("image/" + (r && r.type || "png"));
      } else
        return this.getDataURL(r);
    }, o.prototype.convertToPixel = function(r, n, d) {
      return te(this, "convertToPixel", r, n, d);
    }, o.prototype.convertToLayout = function(r, n, d) {
      return te(this, "convertToLayout", r, n, d);
    }, o.prototype.convertFromPixel = function(r, n, d) {
      return te(this, "convertFromPixel", r, n, d);
    }, o.prototype.containPixel = function(r, n) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      var d = this._model, s, g = le(d, r);
      return R(g, function(S, E) {
        E.indexOf("Models") >= 0 && R(S, function(m) {
          var y = m.coordinateSystem;
          if (y && y.containPoint)
            s = s || !!y.containPoint(n);
          else if (E === "seriesModels") {
            var e = this._chartsMap[m.__viewId];
            e && e.containPoint ? s = s || e.containPoint(n, m) : process.env.NODE_ENV !== "production" && b(E + ": " + (e ? "The found component do not support containPoint." : "No view mapping to the found component."));
          } else
            process.env.NODE_ENV !== "production" && b(E + ": containPoint is not supported");
        }, this);
      }, this), !!s;
    }, o.prototype.getVisual = function(r, n) {
      var d = this._model, s = le(d, r, {
        defaultMainType: "series"
      }), g = s.seriesModel;
      process.env.NODE_ENV !== "production" && (g || b("There is no specified series model"));
      var S = g.getData(), E = s.hasOwnProperty("dataIndexInside") ? s.dataIndexInside : s.hasOwnProperty("dataIndex") ? S.indexOfRawIndex(s.dataIndex) : null;
      return E != null ? ur(S, E, n) : fr(S, n);
    }, o.prototype.getViewOfComponentModel = function(r) {
      return this._componentsMap[r.__viewId];
    }, o.prototype.getViewOfSeriesModel = function(r) {
      return this._chartsMap[r.__viewId];
    }, o.prototype._initEvents = function() {
      var r = this;
      R(xr, function(d) {
        var s = function(g) {
          var S = r.getModel(), E = g.target, m, y = d === "globalout";
          if (y ? m = {} : E && J(E, function(u) {
            var p = $(u);
            if (p && p.dataIndex != null) {
              var f = p.dataModel || S.getSeriesByIndex(p.seriesIndex);
              return m = f && f.getDataParams(p.dataIndex, p.dataType, E) || {}, !0;
            } else if (p.eventData)
              return m = z({}, p.eventData), !0;
          }, !0), m) {
            var e = m.componentType, i = m.componentIndex;
            (e === "markLine" || e === "markPoint" || e === "markArea") && (e = "series", i = m.seriesIndex);
            var t = e && i != null && S.getComponent(e, i), a = t && r[t.mainType === "series" ? "_chartsMap" : "_componentsMap"][t.__viewId];
            process.env.NODE_ENV !== "production" && !y && !(t && a) && b("model or view can not be found by params"), m.event = g, m.type = d, r._$eventProcessor.eventInfo = {
              targetEl: E,
              packedEvent: m,
              model: t,
              view: a
            }, r.trigger(d, m);
          }
        };
        s.zrEventfulCallAtLast = !0, r._zr.on(d, s, r);
      });
      var n = this._messageCenter;
      R(Se, function(d, s) {
        n.on(s, function(g) {
          r.trigger(s, g);
        });
      }), lr(n, this, this._api);
    }, o.prototype.isDisposed = function() {
      return this._disposed;
    }, o.prototype.clear = function() {
      if (this._disposed) {
        D(this.id);
        return;
      }
      this.setOption({
        series: []
      }, !0);
    }, o.prototype.dispose = function() {
      if (this._disposed) {
        D(this.id);
        return;
      }
      this._disposed = !0;
      var r = this.getDom();
      r && Kt(this.getDom(), Ur, "");
      var n = this, d = n._api, s = n._model;
      R(n._componentsViews, function(g) {
        g.dispose(s, d);
      }), R(n._chartsViews, function(g) {
        g.dispose(s, d);
      }), n._zr.dispose(), n._dom = n._model = n._chartsMap = n._componentsMap = n._chartsViews = n._componentsViews = n._scheduler = n._api = n._zr = n._throttledZrFlush = n._theme = n._coordSysMgr = n._messageCenter = null, delete rt[n.id];
    }, o.prototype.resize = function(r) {
      if (this[C]) {
        process.env.NODE_ENV !== "production" && ie("`resize` should not be called during main process.");
        return;
      }
      if (this._disposed) {
        D(this.id);
        return;
      }
      this._zr.resize(r);
      var n = this._model;
      if (this._loadingFX && this._loadingFX.resize(), !!n) {
        var d = n.resetOption("media"), s = r && r.silent;
        this[O] && (s == null && (s = this[O].silent), d = !0, this[O] = null), this[C] = !0, H(this);
        try {
          d && k(this), x.update.call(this, {
            type: "resize",
            animation: z({
              // Disable animation
              duration: 0
            }, r && r.animation)
          });
        } catch (g) {
          throw this[C] = !1, g;
        }
        this[C] = !1, F.call(this, s), Y.call(this, s);
      }
    }, o.prototype.showLoading = function(r, n) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      if (B(r) && (n = r, r = ""), r = r || "default", this.hideLoading(), !Ce[r]) {
        process.env.NODE_ENV !== "production" && b("Loading effects " + r + " not exists.");
        return;
      }
      var d = Ce[r](this._api, n), s = this._zr;
      this._loadingFX = d, s.add(d);
    }, o.prototype.hideLoading = function() {
      if (this._disposed) {
        D(this.id);
        return;
      }
      this._loadingFX && this._zr.remove(this._loadingFX), this._loadingFX = null;
    }, o.prototype.makeActionFromEvent = function(r) {
      var n = z({}, r);
      return n.type = Te[r.type], n;
    }, o.prototype.dispatchAction = function(r, n) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      if (B(n) || (n = {
        silent: !!n
      }), !!ne[r.type] && this._model) {
        if (this[C]) {
          this._pendingActions.push(r);
          return;
        }
        var d = n.silent;
        ge.call(this, r, d);
        var s = n.flush;
        s ? this._zr.flush() : s !== !1 && q.browser.weChat && this._throttledZrFlush(), F.call(this, d), Y.call(this, d);
      }
    }, o.prototype.updateLabelLayout = function() {
      L.trigger("series:layoutlabels", this._model, this._api, {
        // Not adding series labels.
        // TODO
        updatedSeries: []
      });
    }, o.prototype.appendData = function(r) {
      if (this._disposed) {
        D(this.id);
        return;
      }
      var n = r.seriesIndex, d = this.getModel(), s = d.getSeriesByIndex(n);
      process.env.NODE_ENV !== "production" && G(r.data && s), s.appendData(r), this._scheduler.unfinished = !0, this.getZr().wakeUp();
    }, o.internalField = (function() {
      k = function(e) {
        var i = e._scheduler;
        i.restorePipelines(e._model), i.prepareStageTasks(), ve(e, !0), ve(e, !1), i.plan();
      }, ve = function(e, i) {
        for (var t = e._model, a = e._scheduler, u = i ? e._componentsViews : e._chartsViews, p = i ? e._componentsMap : e._chartsMap, f = e._zr, h = e._api, l = 0; l < u.length; l++)
          u[l].__alive = !1;
        i ? t.eachComponent(function(_, A) {
          _ !== "series" && v(A);
        }) : t.eachSeries(v);
        function v(_) {
          var A = _.__requireNewView;
          _.__requireNewView = !1;
          var P = "_ec_" + _.id + "_" + _.type, w = !A && p[P];
          if (!w) {
            var M = Ke(_.type), N = i ? Vt.getClass(M.main, M.sub) : (
              // FIXME:TS
              // (ChartView as ChartViewConstructor).getClass('series', classType.sub)
              // For backward compat, still support a chart type declared as only subType
              // like "liquidfill", but recommend "series.liquidfill"
              // But need a base class to make a type series.
              de.getClass(M.sub)
            );
            process.env.NODE_ENV !== "production" && G(N, M.sub + " does not exist."), w = new N(), w.init(t, h), p[P] = w, u.push(w), f.add(w.group);
          }
          _.__viewId = w.__id = P, w.__alive = !0, w.__model = _, w.group.__ecComponentInfo = {
            mainType: _.mainType,
            index: _.componentIndex
          }, !i && a.prepareView(w, _, t, h);
        }
        for (var l = 0; l < u.length; ) {
          var T = u[l];
          T.__alive ? l++ : (!i && T.renderTask.dispose(), f.remove(T.group), T.dispose(t, h), u.splice(l, 1), p[T.__id] === T && delete p[T.__id], T.__id = T.group.__ecComponentInfo = null);
        }
      }, ee = function(e, i, t, a, u) {
        var p = e._model;
        if (p.setUpdatePayload(t), !a) {
          R([].concat(e._componentsViews).concat(e._chartsViews), T);
          return;
        }
        var f = {};
        f[a + "Id"] = t[a + "Id"], f[a + "Index"] = t[a + "Index"], f[a + "Name"] = t[a + "Name"];
        var h = {
          mainType: a,
          query: f
        };
        u && (h.subType = u);
        var l = t.excludeSeriesId, v;
        l != null && (v = ue(), R(Jt(l), function(_) {
          var A = $t(_, null);
          A != null && v.set(A, !0);
        })), p && p.eachComponent(h, function(_) {
          var A = v && v.get(_.id) != null;
          if (!A)
            if (Ye(t))
              if (_ instanceof ze)
                t.type === Z && !t.notBlur && !_.get(["emphasis", "disabled"]) && Zt(_, t, e._api);
              else {
                var P = Wt(_.mainType, _.componentIndex, t.name, e._api), w = P.focusSelf, M = P.dispatchers;
                t.type === Z && w && !t.notBlur && Xt(_.mainType, _.componentIndex, e._api), M && R(M, function(N) {
                  t.type === Z ? Be(N) : Ge(N);
                });
              }
            else Fe(t) && _ instanceof ze && (qt(_, t, e._api), He(_), V(e));
        }, e), p && p.eachComponent(h, function(_) {
          var A = v && v.get(_.id) != null;
          A || T(e[a === "series" ? "_chartsMap" : "_componentsMap"][_.__viewId]);
        }, e);
        function T(_) {
          _ && _.__alive && _[i] && _[i](_.__model, p, e._api, t);
        }
      }, x = {
        prepareAndUpdate: function(e) {
          k(this), x.update.call(this, e, e && {
            // Needs to mark option changed if newOption is given.
            // It's from MagicType.
            // TODO If use a separate flag optionChanged in payload?
            optionChanged: e.newOption != null
          });
        },
        update: function(e, i) {
          var t = this._model, a = this._api, u = this._zr, p = this._coordSysMgr, f = this._scheduler;
          if (t) {
            t.setUpdatePayload(e), f.restoreData(t, e), f.performSeriesTasks(t), p.create(t, a), f.performDataProcessorTasks(t, e), _e(this, t), p.update(t, a), n(t), f.performVisualTasks(t, e);
            var h = t.get("backgroundColor") || "transparent";
            u.setBackgroundColor(h);
            var l = t.get("darkMode");
            l != null && l !== "auto" && u.setDarkMode(l), me(this, t, a, e, i), L.trigger("afterupdate", t, a);
          }
        },
        updateTransform: function(e) {
          var i = this, t = this._model, a = this._api;
          if (t) {
            t.setUpdatePayload(e);
            var u = [];
            t.eachComponent(function(f, h) {
              if (f !== "series") {
                var l = i.getViewOfComponentModel(h);
                if (l && l.__alive)
                  if (l.updateTransform) {
                    var v = l.updateTransform(h, t, a, e);
                    v && v.update && u.push(l);
                  } else
                    u.push(l);
              }
            });
            var p = ue();
            t.eachSeries(function(f) {
              var h = i._chartsMap[f.__viewId];
              if (h.updateTransform) {
                var l = h.updateTransform(f, t, a, e);
                l && l.update && p.set(f.uid, 1);
              } else
                p.set(f.uid, 1);
            }), n(t), this._scheduler.performVisualTasks(t, e, {
              setDirty: !0,
              dirtyMap: p
            }), re(this, t, a, e, {}, p), L.trigger("afterupdate", t, a);
          }
        },
        updateView: function(e) {
          var i = this._model;
          i && (i.setUpdatePayload(e), de.markUpdateMethod(e, "updateView"), n(i), this._scheduler.performVisualTasks(i, e, {
            setDirty: !0
          }), me(this, i, this._api, e, {}), L.trigger("afterupdate", i, this._api));
        },
        updateVisual: function(e) {
          var i = this, t = this._model;
          t && (t.setUpdatePayload(e), t.eachSeries(function(a) {
            a.getData().clearAllVisual();
          }), de.markUpdateMethod(e, "updateVisual"), n(t), this._scheduler.performVisualTasks(t, e, {
            visualType: "visual",
            setDirty: !0
          }), t.eachComponent(function(a, u) {
            if (a !== "series") {
              var p = i.getViewOfComponentModel(u);
              p && p.__alive && p.updateVisual(u, t, i._api, e);
            }
          }), t.eachSeries(function(a) {
            var u = i._chartsMap[a.__viewId];
            u.updateVisual(a, t, i._api, e);
          }), L.trigger("afterupdate", t, this._api));
        },
        updateLayout: function(e) {
          x.update.call(this, e);
        }
      };
      function r(e, i, t, a, u) {
        if (e._disposed) {
          D(e.id);
          return;
        }
        for (var p = e._model, f = e._coordSysMgr.getCoordinateSystems(), h, l = le(p, t), v = 0; v < f.length; v++) {
          var T = f[v];
          if (T[i] && (h = T[i](p, l, a, u)) != null)
            return h;
        }
        process.env.NODE_ENV !== "production" && b("No coordinate system that supports " + i + " found by the given finder.");
      }
      te = r, _e = function(e, i) {
        var t = e._chartsMap, a = e._scheduler;
        i.eachSeries(function(u) {
          a.updateStreamModes(u, t[u.__viewId]);
        });
      }, ge = function(e, i) {
        var t = this, a = this.getModel(), u = e.type, p = e.escapeConnect, f = ne[u], h = (f.update || "update").split(":"), l = h.pop(), v = h[0] != null && Ke(h[0]);
        this[C] = !0, H(this);
        var T = [e], _ = !1;
        e.batch && (_ = !0, T = Pt(e.batch, function(I) {
          return I = be(z({}, I), e), I.batch = null, I;
        }));
        var A = [], P, w = [], M = f.nonRefinedEventType, N = Fe(e), oe = Ye(e);
        if (oe && xt(this._api), R(T, function(I) {
          var Le = f.action(I, a, t._api);
          if (f.refineEvent ? w.push(Le) : P = Le, P = P || z({}, I), P.type = M, A.push(P), oe) {
            var xe = jt(e), Tt = xe.queryOptionMap, St = xe.mainTypeSpecified, yt = St ? Tt.keys()[0] : "series";
            ee(t, l, I, yt), V(t);
          } else N ? (ee(t, l, I, "series"), V(t)) : v && ee(t, l, I, v.main, v.sub);
        }), l !== "none" && !oe && !N && !v)
          try {
            this[O] ? (k(this), x.update.call(this, e), this[O] = null) : x[l].call(this, e);
          } catch (I) {
            throw this[C] = !1, I;
          }
        if (_ ? P = {
          type: M,
          escapeConnect: p,
          batch: A
        } : P = A[0], this[C] = !1, !i) {
          var U = void 0;
          if (f.refineEvent) {
            var Ve = f.refineEvent(w, e, a, this._api).eventContent;
            G(B(Ve)), U = be({
              type: f.refinedEventType
            }, Ve), U.fromAction = e.type, U.fromActionPayload = e, U.escapeConnect = !0;
          }
          var Me = this._messageCenter;
          Me.trigger(P.type, P), U && Me.trigger(U.type, U);
        }
      }, F = function(e) {
        for (var i = this._pendingActions; i.length; ) {
          var t = i.shift();
          ge.call(this, t, e);
        }
      }, Y = function(e) {
        !e && this.trigger("updated");
      }, $e = function(e, i) {
        e.on("rendered", function(t) {
          i.trigger("rendered", t), // Although zr is dirty if initial animation is not finished
          // and this checking is called on frame, we also check
          // animation finished for robustness.
          e.animation.isFinished() && !i[O] && !i._scheduler.unfinished && !i._pendingActions.length && i.trigger("finished");
        });
      }, Je = function(e, i) {
        e.on("mouseover", function(t) {
          var a = t.target, u = J(a, Ze);
          u && (Ht(u, t, i._api), V(i));
        }).on("mouseout", function(t) {
          var a = t.target, u = J(a, Ze);
          u && (Bt(u, t, i._api), V(i));
        }).on("click", function(t) {
          var a = t.target, u = J(a, function(h) {
            return $(h).dataIndex != null;
          }, !0);
          if (u) {
            var p = u.selected ? "unselect" : "select", f = $(u);
            i._api.dispatchAction({
              type: p,
              dataType: f.dataType,
              dataIndexInside: f.dataIndex,
              seriesIndex: f.seriesIndex,
              isFromClick: !0
            });
          }
        });
      };
      function n(e) {
        e.clearColorPalette(), e.eachSeries(function(i) {
          i.clearColorPalette();
        });
      }
      function d(e) {
        var i = [], t = [], a = !1;
        if (e.eachComponent(function(h, l) {
          var v = l.get("zlevel") || 0, T = l.get("z") || 0, _ = l.getZLevelKey();
          a = a || !!_, (h === "series" ? t : i).push({
            zlevel: v,
            z: T,
            idx: l.componentIndex,
            type: h,
            key: _
          });
        }), a) {
          var u = i.concat(t), p, f;
          fe(u, function(h, l) {
            return h.zlevel === l.zlevel ? h.z - l.z : h.zlevel - l.zlevel;
          }), R(u, function(h) {
            var l = e.getComponent(h.type, h.idx), v = h.zlevel, T = h.key;
            p != null && (v = Math.max(p, v)), T ? (v === p && T !== f && v++, f = T) : f && (v === p && v++, f = ""), p = v, l.setZLevel(v);
          });
        }
      }
      me = function(e, i, t, a, u) {
        d(i), Qe(e, i, t, a, u), R(e._chartsViews, function(p) {
          p.__alive = !1;
        }), re(e, i, t, a, u), R(e._chartsViews, function(p) {
          p.__alive || p.remove(i, t);
        });
      }, Qe = function(e, i, t, a, u, p) {
        R(p || e._componentsViews, function(f) {
          var h = f.__model;
          m(h, f), f.render(h, i, t, a), E(h, f), y(h, f);
        });
      }, re = function(e, i, t, a, u, p) {
        var f = e._scheduler;
        u = z(u || {}, {
          updatedSeries: i.getSeries()
        }), L.trigger("series:beforeupdate", i, t, u);
        var h = !1;
        i.eachSeries(function(l) {
          var v = e._chartsMap[l.__viewId];
          v.__alive = !0;
          var T = v.renderTask;
          f.updatePayload(T, a), m(l, v), p && p.get(l.uid) && T.dirty(), T.perform(f.getPerformArgs(T)) && (h = !0), v.group.silent = !!l.get("silent"), S(l, v), He(l);
        }), f.unfinished = h || f.unfinished, L.trigger("series:layoutlabels", i, t, u), L.trigger("series:transition", i, t, u), i.eachSeries(function(l) {
          var v = e._chartsMap[l.__viewId];
          E(l, v), y(l, v);
        }), g(e, i), L.trigger("series:afterupdate", i, t, u);
      }, V = function(e) {
        e[he] = !0, e.getZr().wakeUp();
      }, H = function(e) {
        e[Q] = (e[Q] + 1) % 1e3;
      }, tt = function(e) {
        e[he] && (e.getZr().storage.traverse(function(i) {
          ce(i) || s(i);
        }), e[he] = !1);
      };
      function s(e) {
        for (var i = [], t = e.currentStates, a = 0; a < t.length; a++) {
          var u = t[a];
          u === "emphasis" || u === "blur" || u === "select" || i.push(u);
        }
        e.selected && e.states.select && i.push("select"), e.hoverState === Nt && e.states.emphasis ? i.push("emphasis") : e.hoverState === Ut && e.states.blur && i.push("blur"), e.useStates(i);
      }
      function g(e, i) {
        var t = e._zr, a = t.storage, u = 0;
        a.traverse(function(p) {
          p.isGroup || u++;
        }), u > i.get("hoverLayerThreshold") && !q.node && !q.worker && i.eachSeries(function(p) {
          if (!p.preventUsingHoverLayer) {
            var f = e._chartsMap[p.__viewId];
            f.__alive && f.eachRendered(function(h) {
              h.states.emphasis && (h.states.emphasis.hoverLayer = !0);
            });
          }
        });
      }
      function S(e, i) {
        var t = e.get("blendMode") || null;
        i.eachRendered(function(a) {
          a.isGroup || (a.style.blend = t);
        });
      }
      function E(e, i) {
        if (!e.preventAutoZ) {
          var t = Mt(e);
          i.eachRendered(function(a) {
            return Lt(a, t.z, t.zlevel), !0;
          });
        }
      }
      function m(e, i) {
        i.eachRendered(function(t) {
          if (!ce(t)) {
            var a = t.getTextContent(), u = t.getTextGuideLine();
            t.stateTransition && (t.stateTransition = null), a && a.stateTransition && (a.stateTransition = null), u && u.stateTransition && (u.stateTransition = null), t.hasState() ? (t.prevStates = t.currentStates, t.clearStates()) : t.prevStates && (t.prevStates = null);
          }
        });
      }
      function y(e, i) {
        var t = e.getModel("stateAnimation"), a = e.isAnimationEnabled(), u = t.get("duration"), p = u > 0 ? {
          duration: u,
          delay: t.get("delay"),
          easing: t.get("easing")
          // additive: stateAnimationModel.get('additive')
        } : null;
        i.eachRendered(function(f) {
          if (f.states && f.states.emphasis) {
            if (ce(f))
              return;
            if (f instanceof Tr && kt(f), f.__dirty) {
              var h = f.prevStates;
              h && f.useStates(h);
            }
            if (a) {
              f.stateTransition = p;
              var l = f.getTextContent(), v = f.getTextGuideLine();
              l && (l.stateTransition = p), v && (v.stateTransition = p);
            }
            f.__dirty && s(f);
          }
        });
      }
      et = function(e) {
        return new /** @class */
        ((function(i) {
          Ee(t, i);
          function t() {
            return i !== null && i.apply(this, arguments) || this;
          }
          return t.prototype.getCoordinateSystems = function() {
            return e._coordSysMgr.getCoordinateSystems();
          }, t.prototype.getComponentByElement = function(a) {
            for (; a; ) {
              var u = a.__ecComponentInfo;
              if (u != null)
                return e._model.getComponent(u.mainType, u.index);
              a = a.parent;
            }
          }, t.prototype.enterEmphasis = function(a, u) {
            Be(a, u), V(e);
          }, t.prototype.leaveEmphasis = function(a, u) {
            Ge(a, u), V(e);
          }, t.prototype.enterBlur = function(a) {
            bt(a), V(e);
          }, t.prototype.leaveBlur = function(a) {
            zt(a), V(e);
          }, t.prototype.enterSelect = function(a) {
            Ft(a), V(e);
          }, t.prototype.leaveSelect = function(a) {
            Yt(a), V(e);
          }, t.prototype.getModel = function() {
            return e.getModel();
          }, t.prototype.getViewOfComponentModel = function(a) {
            return e.getViewOfComponentModel(a);
          }, t.prototype.getViewOfSeriesModel = function(a) {
            return e.getViewOfSeriesModel(a);
          }, t.prototype.getMainProcessVersion = function() {
            return e[Q];
          }, t;
        })(At))(e);
      };
    })(), o;
  })(Oe)
), Ae = Lr.prototype;
Ae.on = lt("on");
Ae.off = lt("off");
Ae.one = function(c, o, r) {
  var n = this;
  dr("ECharts#one is deprecated.");
  function d() {
    for (var s = [], g = 0; g < arguments.length; g++)
      s[g] = arguments[g];
    o && o.apply && o.apply(this, s), n.off(c, d);
  }
  this.on.call(this, c, d, r);
};
var xr = ["click", "dblclick", "mouseover", "mouseout", "mousemove", "mousedown", "mouseup", "globalout", "contextmenu"];
function D(c) {
  process.env.NODE_ENV !== "production" && b("Instance " + c + " has been disposed");
}
var ne = {}, Te = {}, Se = {}, ye = [], Re = [], ae = [], gt = {}, Ce = {}, rt = {}, Nr = {}, Ur = "_echarts_instance_";
function mt(c, o) {
  gt[c] = o;
}
function kr(c) {
  nt(Re, c) < 0 && Re.push(c);
}
function br(c, o) {
  De(ye, c, o, Pr);
}
function Ai(c) {
  Et("afterinit", c);
}
function Di(c) {
  Et("afterupdate", c);
}
function Et(c, o) {
  L.on(c, o);
}
function j(c, o, r) {
  var n, d, s, g, S;
  at(o) && (r = o, o = ""), B(c) ? (n = c.type, d = c.event, g = c.update, S = c.publishNonRefinedEvent, r || (r = c.action), s = c.refineEvent) : (n = c, d = o);
  function E(y) {
    return y.toLowerCase();
  }
  d = E(d || n);
  var m = s ? E(n) : d;
  ne[n] || (G(je.test(n) && je.test(d)), s && G(d !== n), ne[n] = {
    actionType: n,
    refinedEventType: d,
    nonRefinedEventType: m,
    update: g,
    action: r,
    refineEvent: s
  }, Se[d] = 1, s && S && (Se[m] = 1), process.env.NODE_ENV !== "production" && Te[m] && ie(m + ' must not be shared; use "refineEvent" if you intend to share an event name.'), Te[m] = n);
}
function Ii(c, o) {
  st.register(c, o);
}
function Vi(c, o) {
  De(ae, c, o, ft, "layout");
}
function W(c, o) {
  De(ae, c, o, dt, "visual");
}
var it = [];
function De(c, o, r, n, d) {
  if ((at(o) || B(o)) && (r = o, o = n), process.env.NODE_ENV !== "production") {
    if (isNaN(o) || o == null)
      throw new Error("Illegal priority");
    R(c, function(g) {
      G(g.__raw !== r);
    });
  }
  if (!(nt(it, r) >= 0)) {
    it.push(r);
    var s = ut.wrapStageHandler(r, d);
    s.__prio = o, s.__raw = r, c.push(s);
  }
}
function zr(c, o) {
  Ce[c] = o;
}
function Mi(c, o, r) {
  var n = mr("registerMap");
  n && n(c, o, r);
}
var Li = cr;
W(we, er);
W(se, tr);
W(se, rr);
W(we, sr);
W(se, or);
W(pt, _r);
kr(ot);
br(Cr, It);
zr("default", ir);
j({
  type: Z,
  event: Z,
  update: Z
}, K);
j({
  type: pe,
  event: pe,
  update: pe
}, K);
j({
  type: We,
  event: Pe,
  update: We,
  action: K,
  refineEvent: Ie,
  publishNonRefinedEvent: !0
});
j({
  type: Xe,
  event: Pe,
  update: Xe,
  action: K,
  refineEvent: Ie,
  publishNonRefinedEvent: !0
});
j({
  type: qe,
  event: Pe,
  update: qe,
  action: K,
  refineEvent: Ie,
  publishNonRefinedEvent: !0
});
function Ie(c, o, r, n) {
  return {
    eventContent: {
      selected: Gt(r),
      isFromClick: o.isFromClick || !1
    }
  };
}
mt("default", {});
mt("dark", nr);
export {
  wi as PRIORITY,
  j as registerAction,
  Ii as registerCoordinateSystem,
  Vi as registerLayout,
  zr as registerLoading,
  Ui as registerLocale,
  Mi as registerMap,
  Ai as registerPostInit,
  Di as registerPostUpdate,
  kr as registerPreprocessor,
  br as registerProcessor,
  mt as registerTheme,
  Li as registerTransform,
  Et as registerUpdateLifecycle,
  W as registerVisual
};
