// echarts/lib/util/model.js
import { createHashMap as M, each as v, isObject as y, isString as w, isNumber as D, isStringSafe as k, indexOf as O, assert as q, isArray as E, map as T } from "../../../zrender/lib/core/util.js";
import X from "../../../zrender/lib/core/env.js";
import { isNumeric as Y, getRandomIdBase as G, round as m, getPrecision as g } from "./number.js";
import { error as p, warn as K } from "./log.js";
function S(n, e, r) {
  return (e - n) * r + n;
}
var P = "series\0", B = "\0_ec_\0";
function U(n) {
  return n instanceof Array ? n : n == null ? [] : [n];
}
function ln(n, e, r) {
  if (n) {
    n[e] = n[e] || {}, n.emphasis = n.emphasis || {}, n.emphasis[e] = n.emphasis[e] || {};
    for (var i = 0, a = r.length; i < a; i++) {
      var u = r[i];
      !n.emphasis[e].hasOwnProperty(u) && n[e].hasOwnProperty(u) && (n.emphasis[e][u] = n[e][u]);
    }
  }
}
var tn = ["fontStyle", "fontWeight", "fontSize", "fontFamily", "rich", "tag", "color", "textBorderColor", "textBorderWidth", "width", "height", "lineHeight", "align", "verticalAlign", "baseline", "shadowColor", "shadowBlur", "shadowOffsetX", "shadowOffsetY", "textShadowColor", "textShadowBlur", "textShadowOffsetX", "textShadowOffsetY", "backgroundColor", "borderColor", "borderWidth", "borderRadius", "padding"];
function fn(n) {
  return y(n) && !E(n) && !(n instanceof Date) ? n.value : n;
}
function on(n) {
  return y(n) && !(n instanceof Array);
}
function sn(n, e, r) {
  var i = r === "normalMerge", a = r === "replaceMerge", u = r === "replaceAll";
  n = n || [], e = (e || []).slice();
  var t = M();
  v(e, function(l, o) {
    if (!y(l)) {
      e[o] = null;
      return;
    }
    process.env.NODE_ENV !== "production" && (l.id != null && !A(l.id) && R(l.id), l.name != null && !A(l.name) && R(l.name));
  });
  var f = W(n, t, r);
  return (i || a) && z(f, n, t, e), i && C(f, e), i || a ? H(f, e, a) : u && $(f, e), J(f), f;
}
function W(n, e, r) {
  var i = [];
  if (r === "replaceAll")
    return i;
  for (var a = 0; a < n.length; a++) {
    var u = n[a];
    u && u.id != null && e.set(u.id, a), i.push({
      existing: r === "replaceMerge" || _(u) ? null : u,
      newOption: null,
      keyInfo: null,
      brandNew: null
    });
  }
  return i;
}
function z(n, e, r, i) {
  v(i, function(a, u) {
    if (!(!a || a.id == null)) {
      var t = I(a.id), f = r.get(t);
      if (f != null) {
        var l = n[f];
        q(!l.newOption, 'Duplicated option on id "' + t + '".'), l.newOption = a, l.existing = e[f], i[u] = null;
      }
    }
  });
}
function C(n, e) {
  v(e, function(r, i) {
    if (!(!r || r.name == null))
      for (var a = 0; a < n.length; a++) {
        var u = n[a].existing;
        if (!n[a].newOption && u && (u.id == null || r.id == null) && !_(r) && !_(u) && L("name", u, r)) {
          n[a].newOption = r, e[i] = null;
          return;
        }
      }
  });
}
function H(n, e, r) {
  v(e, function(i) {
    if (i) {
      for (
        var a, u = 0;
        // Be `!resultItem` only when `nextIdx >= result.length`.
        (a = n[u]) && (a.newOption || _(a.existing) || // In mode "replaceMerge", here no not-mapped-non-internal-existing.
        a.existing && i.id != null && !L("id", i, a.existing));
      )
        u++;
      a ? (a.newOption = i, a.brandNew = r) : n.push({
        newOption: i,
        brandNew: r,
        existing: null,
        keyInfo: null
      }), u++;
    }
  });
}
function $(n, e) {
  v(e, function(r) {
    n.push({
      newOption: r,
      brandNew: !0,
      existing: null,
      keyInfo: null
    });
  });
}
function J(n) {
  var e = M();
  v(n, function(r) {
    var i = r.existing;
    i && e.set(i.id, r);
  }), v(n, function(r) {
    var i = r.newOption;
    q(!i || i.id == null || !e.get(i.id) || e.get(i.id) === r, "id duplicates: " + (i && i.id)), i && i.id != null && e.set(i.id, r), !r.keyInfo && (r.keyInfo = {});
  }), v(n, function(r, i) {
    var a = r.existing, u = r.newOption, t = r.keyInfo;
    if (y(u)) {
      if (t.name = u.name != null ? I(u.name) : a ? a.name : P + i, a)
        t.id = I(a.id);
      else if (u.id != null)
        t.id = I(u.id);
      else {
        var f = 0;
        do
          t.id = "\0" + t.name + "\0" + f++;
        while (e.get(t.id));
      }
      e.set(t.id, r);
    }
  });
}
function L(n, e, r) {
  var i = N(e[n], null), a = N(r[n], null);
  return i != null && a != null && i === a;
}
function I(n) {
  if (process.env.NODE_ENV !== "production" && n == null)
    throw new Error();
  return N(n, "");
}
function N(n, e) {
  return n == null ? e : w(n) ? n : D(n) || k(n) ? n + "" : e;
}
function R(n) {
  process.env.NODE_ENV !== "production" && K("`" + n + "` is invalid id or name. Must be a string or number.");
}
function A(n) {
  return k(n) || Y(n);
}
function dn(n) {
  var e = n.name;
  return !!(e && e.indexOf(P));
}
function _(n) {
  return n && n.id != null && I(n.id).indexOf(B) === 0;
}
function vn(n) {
  return B + n;
}
function cn(n, e, r) {
  v(n, function(i) {
    var a = i.newOption;
    y(a) && (i.keyInfo.mainType = e, i.keyInfo.subType = Q(e, a, i.existing, r));
  });
}
function Q(n, e, r, i) {
  var a = e.type ? e.type : r ? r.subType : i.determineSubType(n, e);
  return a;
}
function hn(n, e) {
  var r = {}, i = {};
  return a(n || [], r), a(e || [], i, r), [u(r), u(i)];
  function a(t, f, l) {
    for (var o = 0, d = t.length; o < d; o++) {
      var s = N(t[o].seriesId, null);
      if (s == null)
        return;
      for (var c = U(t[o].dataIndex), h = l && l[s], x = 0, F = c.length; x < F; x++) {
        var b = c[x];
        h && h[b] ? h[b] = null : (f[s] || (f[s] = {}))[b] = 1;
      }
    }
  }
  function u(t, f) {
    var l = [];
    for (var o in t)
      if (t.hasOwnProperty(o) && t[o] != null)
        if (f)
          l.push(+o);
        else {
          var d = u(t[o], !0);
          d.length && l.push({
            seriesId: o,
            dataIndex: d
          });
        }
    return l;
  }
}
function xn(n, e) {
  if (e.dataIndexInside != null)
    return e.dataIndexInside;
  if (e.dataIndex != null)
    return E(e.dataIndex) ? T(e.dataIndex, function(r) {
      return n.indexOfRawIndex(r);
    }) : n.indexOfRawIndex(e.dataIndex);
  if (e.name != null)
    return E(e.name) ? T(e.name, function(r) {
      return n.indexOfName(r);
    }) : n.indexOfName(e.name);
}
function In() {
  var n = "__ec_inner_" + Z++;
  return function(e) {
    return e[n] || (e[n] = {});
  };
}
var Z = G();
function yn(n, e, r) {
  var i = j(e, r), a = i.mainTypeSpecified, u = i.queryOptionMap, t = i.others, f = t, l = r ? r.defaultMainType : null;
  return !a && l && u.set(l, {}), u.each(function(o, d) {
    var s = nn(n, d, o, {
      useDefault: l === d,
      enableAll: r && r.enableAll != null ? r.enableAll : !0,
      enableNone: r && r.enableNone != null ? r.enableNone : !0
    });
    f[d + "Models"] = s.models, f[d + "Model"] = s.models[0];
  }), f;
}
function j(n, e) {
  var r;
  if (w(n)) {
    var i = {};
    i[n + "Index"] = 0, r = i;
  } else
    r = n;
  var a = M(), u = {}, t = !1;
  return v(r, function(f, l) {
    if (l === "dataIndex" || l === "dataIndexInside") {
      u[l] = f;
      return;
    }
    var o = l.match(/^(\w+)(Index|Id|Name)$/) || [], d = o[1], s = (o[2] || "").toLowerCase();
    if (!(!d || !s || e && e.includeMainTypes && O(e.includeMainTypes, d) < 0)) {
      t = t || !!d;
      var c = a.get(d) || a.set(d, {});
      c[s] = f;
    }
  }), {
    mainTypeSpecified: t,
    queryOptionMap: a,
    others: u
  };
}
var V = {
  useDefault: !0,
  enableAll: !1,
  enableNone: !1
}, gn = {
  useDefault: !1,
  enableAll: !0,
  enableNone: !0
};
function nn(n, e, r, i) {
  i = i || V;
  var a = r.index, u = r.id, t = r.name, f = {
    models: null,
    specified: a != null || u != null || t != null
  };
  if (!f.specified) {
    var l = void 0;
    return f.models = i.useDefault && (l = n.getComponent(e)) ? [l] : [], f;
  }
  if (a === "none" || a === !1) {
    if (i.enableNone)
      return f.models = [], f;
    process.env.NODE_ENV !== "production" && p('`"none"` or `false` is not a valid value on index option.'), a = -1;
  }
  return a === "all" && (i.enableAll ? a = u = t = null : (process.env.NODE_ENV !== "production" && p('`"all"` is not a valid value on index option.'), a = -1)), f.models = n.queryComponents({
    mainType: e,
    index: a,
    id: u,
    name: t
  }), f;
}
function Nn(n, e, r) {
  n.setAttribute ? n.setAttribute(e, r) : n[e] = r;
}
function _n(n) {
  return n === "auto" ? X.domSupported ? "html" : "richText" : n || "html";
}
function Mn(n, e) {
  var r = M(), i = [];
  return v(n, function(a) {
    var u = e(a);
    (r.get(u) || (i.push(u), r.set(u, []))).push(a);
  }), {
    keys: i,
    buckets: r
  };
}
function bn(n, e, r, i, a) {
  var u = e == null || e === "auto";
  if (i == null)
    return i;
  if (D(i)) {
    var t = S(r || 0, i, a);
    return m(t, u ? Math.max(g(r || 0), g(i)) : e);
  } else {
    if (w(i))
      return a < 1 ? r : i;
    for (var f = [], l = r, o = i, d = Math.max(l ? l.length : 0, o.length), s = 0; s < d; ++s) {
      var c = n.getDimensionInfo(s);
      if (c && c.type === "ordinal")
        f[s] = (a < 1 && l ? l : o)[s];
      else {
        var h = l && l[s] ? l[s] : 0, x = o[s], t = S(h, x, a);
        f[s] = m(t, u ? Math.max(g(h), g(x)) : e);
      }
    }
    return f;
  }
}
var En = (
  /** @class */
  (function() {
    function n() {
    }
    return n.prototype.reset = function(e, r, i, a) {
      return this._list = e, this._step = a = a || 1, this._idx = r, this._end = i ?? (a > 0 ? e.length : 0), this.item = null, this.key = NaN, this;
    }, n.prototype.next = function() {
      return (this._step > 0 ? this._idx < this._end : this._idx >= this._end) ? (this.item = this._list[this._idx], this.key = this._idx = this._idx + this._step, !0) : !1;
    }, n;
  })()
);
function wn(n) {
  n.option = n.parentModel = n.ecModel = null;
}
export {
  En as ListIterator,
  gn as MULTIPLE_REFERRING,
  V as SINGLE_REFERRING,
  tn as TEXT_STYLE_OPTIONS,
  wn as clearTmpModel,
  hn as compressBatches,
  N as convertOptionIdName,
  ln as defaultEmphasis,
  fn as getDataItemValue,
  _n as getTooltipRenderMode,
  Mn as groupData,
  bn as interpolateRawValues,
  _ as isComponentIdInternal,
  on as isDataItemOption,
  dn as isNameSpecified,
  In as makeInner,
  vn as makeInternalComponentId,
  sn as mappingToExists,
  U as normalizeToArray,
  yn as parseFinder,
  j as preParseFinder,
  xn as queryDataIndex,
  nn as queryReferringComponents,
  Nn as setAttribute,
  cn as setComponentTypeToKeyInfo
};
